### PR TITLE
Fix ParticleHandler::sort_particles_into_subdomains_and_cells() for lost particles

### LINKED ISSUE #12237: step-19 crashes after 25 time steps
I ran step-19 and observed a crash, with the following stacktrace:
```
#0  0x00007fffd81953d7 in raise () from /lib64/libc.so.6
#1  0x00007fffd8196ac8 in abort () from /lib64/libc.so.6
#2  0x00007fffec85a471 in dealii::deal_II_exceptions::internals::abort (exc=...)
    at /home/kronbichler/deal/deal.II/source/base/exceptions.cc:489
#3  0x000000000057e45f in dealii::deal_II_exceptions::internals::issue_error_noreturn<dealii::TriaRawIterator<dealii::CellAccessor<2, 2> >::ExcDereferenceInvalidCell> (handling=dealii::deal_II_exceptions::internals::abort_or_throw_on_exception, 
    file=0x5d6768 "/home/kronbichler/deal/deal.II/include/deal.II/grid/tria_iterator.h", line=998, 
    function=0x5dfb80 <dealii::TriaRawIterator<dealii::CellAccessor<2, 2> >::operator*() const::__PRETTY_FUNCTION__> "const Accessor& dealii::TriaRawIterator<Accessor>::operator*() const [with Accessor = dealii::CellAccessor<2, 2>]", 
    cond=0x5da190 "Accessor::structure_dimension != Accessor::dimension || state() == IteratorState::valid", 
    exc_name=0x5da168 "ExcDereferenceInvalidCell(accessor)", e=...)
    at /home/kronbichler/deal/deal.II/include/deal.II/base/exceptions.h:1368
#4  0x0000000000570422 in dealii::TriaRawIterator<dealii::CellAccessor<2, 2> >::operator* (this=0x7fffffff9ee0)
    at /home/kronbichler/deal/deal.II/include/deal.II/grid/tria_iterator.h:996
#5  0x0000000000566d5a in dealii::TriaRawIterator<dealii::CellAccessor<2, 2> >::operator-> (this=0x7fffffff9ee0)
    at /home/kronbichler/deal/deal.II/include/deal.II/grid/tria_iterator.h:1035
#6  0x000000000055d762 in Step19::CathodeRaySimulator<2>::track_lost_particle (this=0x7fffffffc090, particle=..., cell=...)
    at /home/kronbichler/deal/deal.II/examples/step-19/step-19.cc:758
#7  0x000000000055aa16 in Step19::CathodeRaySimulator<2>::CathodeRaySimulator()::{lambda(dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)#1}::operator()(dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&) const (__closure=0x7bdf80, particle=..., cell=...)
    at /home/kronbichler/deal/deal.II/examples/step-19/step-19.cc:214
#8  0x00000000005a362b in boost::detail::function::void_function_obj_invoker2<Step19::CathodeRaySimulator<2>::CathodeRaySimulator()::{lambda(dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)#1}, void, dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&>::invoke(boost::detail::function::function_buffer&, dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&) (function_obj_ptr=..., a0=..., a1=...)
    at /home/kronbichler/deal/deal.II/bundled/boost-1.70.0/include/boost/function/function_template.hpp:158
#9  0x00007ffff17b488c in boost::function2<void, dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&>::operator() (this=0x7bdf78, a0=..., a1=...)
    at /home/kronbichler/deal/deal.II/bundled/boost-1.70.0/include/boost/function/function_template.hpp:763
#10 0x00007ffff17b3116 in boost::signals2::detail::call_with_tuple_args<boost::signals2::detail::void_type>::m_invoke<boost::function<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)>, 0u, 1u, dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&>(boost::function<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)>&, boost::signals2::detail::unsigned_meta_array<0u, 1u>, std::tuple<dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&> const&, boost::enable_if<boost::is_void<boost::function<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)>::result_type>, void>::type*) const (
    this=0x7fffffff99df, func=..., args=std::tuple containing = {...})
    at /home/kronbichler/deal/deal.II/bundled/boost-1.70.0/include/boost/signals2/detail/variadic_slot_invoker.hpp:105
#11 0x00007ffff17b0947 in boost::signals2::detail::call_with_tuple_args<boost::signals2::detail::void_type>::operator()<boost::function<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)>, dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&, 2ul>(boost::function<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)>&, std:---Type <return> to continue, or q <return> to quit---
:tuple<dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&> const&, mpl_::size_t<2ul>) const (this=0x7fffffff99df, func=..., args=std::tuple containing = {...})
    at /home/kronbichler/deal/deal.II/bundled/boost-1.70.0/include/boost/signals2/detail/variadic_slot_invoker.hpp:90
#12 0x00007ffff17abf52 in boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type, dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&>::operator()<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&), boost::function<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)> >, boost::signals2::mutex> > >(boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&), boost::function<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)> >, boost::signals2::mutex> > const&) const (this=0x7fffffff9d70, 
    connectionBody=...)
    at /home/kronbichler/deal/deal.II/bundled/boost-1.70.0/include/boost/signals2/detail/variadic_slot_invoker.hpp:133
#13 0x00007ffff17a565a in boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type, dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&>, std::_List_iterator<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&), boost::function<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)> >, boost::signals2::mutex> > >, boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&), boost::function<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)> >, boost::signals2::mutex> >::dereference() const (this=0x7fffffff9ac0)
    at /home/kronbichler/deal/deal.II/bundled/boost-1.70.0/include/boost/signals2/detail/slot_call_iterator.hpp:110
#14 0x00007ffff179d952 in boost::iterators::iterator_core_access::dereference<boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type, dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&>, std::_List_iterator<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&), boost::function<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)> >, boost::signals2::mutex> > >, boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&), boost::function<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)> >, boost::signals2::mutex> > >(boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type, dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&>, std::_List_iterator<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&), boost::function<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)> >, boost::signals2::mutex> > >, boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&), boost::function<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)> >, boost::signals2::mutex> > const&) (f=...)
---Type <return> to continue, or q <return> to quit---
    at /home/kronbichler/deal/deal.II/bundled/boost-1.70.0/include/boost/iterator/iterator_facade.hpp:550
#15 0x00007ffff179200c in boost::iterators::detail::iterator_facade_base<boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type, dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&>, std::_List_iterator<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&), boost::function<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)> >, boost::signals2::mutex> > >, boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&), boost::function<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)> >, boost::signals2::mutex> >, boost::signals2::detail::void_type, boost::iterators::single_pass_traversal_tag, boost::signals2::detail::void_type const&, long, false, false>::operator*() const (this=0x7fffffff9ac0)
    at /home/kronbichler/deal/deal.II/bundled/boost-1.70.0/include/boost/iterator/iterator_facade.hpp:656
#16 0x00007ffff17826c5 in boost::signals2::optional_last_value<void>::operator()<boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type, dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&>, std::_List_iterator<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&), boost::function<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)> >, boost::signals2::mutex> > >, boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&), boost::function<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)> >, boost::signals2::mutex> > >(boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type, dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&>, std::_List_iterator<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&), boost::function<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)> >, boost::signals2::mutex> > >, boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&), boost::function<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)> >, boost::signals2::mutex> >, boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type, dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&>, std::_List_iterator<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&), boost::function<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)> >, boost::signals2::mutex> > >, boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&), boost::function<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)> >, boost::signals2::mutex> >) const (
    this=0x7bb310, first=..., last=...)
    at /home/kronbichler/deal/deal.II/bundled/boost-1.70.0/include/boost/signals2/optional_last_value.hpp:57
#17 0x00007ffff176c4d0 in boost::signals2::detail::combiner_invoker<void>::operator()<boost::signals2::optional_last_value<void>, boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type, d---Type <return> to continue, or q <return> to quit---
ealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&>, std::_List_iterator<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&), boost::function<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)> >, boost::signals2::mutex> > >, boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&), boost::function<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)> >, boost::signals2::mutex> > >(boost::signals2::optional_last_value<void>&, boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type, dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&>, std::_List_iterator<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&), boost::function<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)> >, boost::signals2::mutex> > >, boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&), boost::function<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)> >, boost::signals2::mutex> >, boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type, dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&>, std::_List_iterator<boost::shared_ptr<boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&), boost::function<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)> >, boost::signals2::mutex> > >, boost::signals2::detail::connection_body<std::pair<boost::signals2::detail::slot_meta_group, boost::optional<int> >, boost::signals2::slot<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&), boost::function<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)> >, boost::signals2::mutex> >) const (this=0x7fffffff9dcf, combiner=..., 
    first=..., last=...)
    at /home/kronbichler/deal/deal.II/bundled/boost-1.70.0/include/boost/signals2/detail/result_type_wrapper.hpp:64
#18 0x00007ffff1751d38 in boost::signals2::detail::signal_impl<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&), boost::signals2::optional_last_value<void>, int, std::less<int>, boost::function<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)>, boost::function<void (boost::signals2::connection const&, dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)>, boost::signals2::mutex>::operator()(dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&) (this=0x7bd8a0, args#0=..., args#1=...)
    at /home/kronbichler/deal/deal.II/bundled/boost-1.70.0/include/boost/signals2/detail/signal_template.hpp:243
#19 0x00007ffff173ae10 in boost::signals2::signal<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&), boost::signals2::optional_last_value<void>, int, std::less<int>, boost::function<void (dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)>, boost::function<void (boost::signals2::connection const&, dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&)>, boost::signals2::mutex>::operator()(dealii::Particles::ParticleIterator<2, 2> const&, dealii::TriaActiveIterator<dealii::CellAccessor<2, 2> > const&) (this=0x7fffffffd2e8, args#0=..., args#1=...)
    at /home/kronbichler/deal/deal.II/bundled/boost-1.70.0/include/boost/signals2/detail/signal_template.hpp:722
#20 0x00007ffff170fa68 in dealii::Particles::ParticleHandler<2, 2>::sort_particles_into_subdomains_and_cells (
    this=0x7fffffffd288) at /home/kronbichler/deal/deal.II/source/particles/particle_handler.cc:1105
---Type <return> to continue, or q <return> to quit---
#21 0x0000000000563f64 in Step19::CathodeRaySimulator<2>::move_particles (this=0x7fffffffc090)
    at /home/kronbichler/deal/deal.II/examples/step-19/step-19.cc:728
#22 0x000000000055af62 in Step19::CathodeRaySimulator<2>::run (this=0x7fffffffc090)
    at /home/kronbichler/deal/deal.II/examples/step-19/step-19.cc:1008
#23 0x000000000054b780 in main () at /home/kronbichler/deal/deal.II/examples/step-19/step-19.cc:1044

```
The interesting part is that it happens when we trace a lost particle. Does anyone know if this failure is recent (e.g. due to #11276 or some follow-up PR), or has happened for a longer time?

### PR DESCRIPTION
**Pull request by peterrum**
As far as I see it this is the correct logic. What the documentation of the signal `particle_lost` states is that the old cell is returned if the particle leaves the particle. This was not the case, since the cell iterator has been overridden...

@gassmoeller May I ask you for a quick review (when you wake up).

closes #12237

### CODE REVIEW

File: source/particles/particle_handler.cc [hunk 1]
```diff
@@ -1067,6 +1067,11 @@ namespace Particles
                                                           cell_centers);
                     });
 
+          // make a copy of the current cell, since we will modify the variable
+          // in the following current_cell but we need a backup in the case
+          // the particle is not found
+          const auto current_cell_old = current_cell;
```
Comment (kronbichler) [inline source/particles/particle_handler.cc:1073]: Shouldn't we call this `previous_cell_of_particle`?

Comment (peterrum) [reply source/particles/particle_handler.cc]: Done.

**Review by kronbichler — APPROVED** [review]
Looks good to me.

**Comment by kronbichler** [comment]
If this fixes #12237, and there appears that we did not have something checking this before, I wonder if there is anything we can add as a test? Maybe step-19 for the first 30 time steps?

File: tests/particles/lost_01.cc [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  particle_handler.insert_particles(particle_positions);
+
+  particle_handler.signals.particle_lost.connect(
+    [](const auto &particle, const auto &cell) {
+      deallog << particle->get_id() << " " << cell->id() << std::endl;
+    });
+
+  particle_handler.begin()->set_location(Point<dim>{2.0, 2.0});
+
+  particle_handler.sort_particles_into_subdomains_and_cells();
```
Comment (peterrum) [inline tests/particles/lost_01.cc:53]: @kronbichler I went with a simpler test. This is all you need to break master^^

Comment (peterrum) [reply tests/particles/lost_01.cc:53]: A single cell and a single particle...

**Review by kronbichler — APPROVED** [review]
Excellent, thank you.

**Review by luca-heltai — APPROVED** [review]

**Comment by gassmoeller** [comment]
This makes sense. I can not quite retrace at what point this signal did or did not hand over the correct cell reference, but the fixed state is definitely the intended state. ASPECT makes no use of this signal, I think step-19 may be the only application for it at the moment, that is how this did not get noticed before. Thanks for fixing @peterrum!

**Pull request merged**